File: Python/BOJ/1780 종이의 개수.py
def div(y,x,n):
    num = paper[y][x]
    for i in range(y,y+n):
        for j in range(x,x+n):
            if num != paper[i][j]:
                m = n//3 # 3분할 해보자
                div(y, x, m) 
                div(y, x + m, m)
                div(y, x + 2*m, m)
                div(y + m, x, m)
                div(y + m, x + m, m)
                div(y + m, x + 2*m, m)
                div(y + 2*m, x, m)
                div(y + 2*m, x + m, m)
                div(y + 2*m, x + 2*m, m) #여기까지 3분할씩 한 영역
                return
    if num == -1:
        result[0] += 1
    elif num == 0:
        result[1] += 1
    else :
        result[2] += 1 

import sys
N = int(sys.stdin.readline())
paper = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]
result = [0,0,0]
div(0,0,N)
for i in range(3):
    print(result[i])

File: Java/BOJ/1780 종이의 개수.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static int[][] paper;
	static int[] result = {0, 0, 0};
	public static void div(int y, int x, int n ) {
		int num = paper[y][x];
		for(int i = y; i < y + n; i++) {
			for (int j = x; j < x + n; j++) {
				if (num != paper[i][j]) {
					int m = n/3;
					div(y, x, m);
					div(y, x + m, m);
					div(y, x + 2*m, m);
					div(y + m, x, m);
					div(y + m, x + m, m);
					div(y + m, x + 2*m, m);
					div(y + 2*m, x, m);
					div(y + 2*m, x + m, m);
					div(y + 2*m, x + 2*m, m);
					return;
				}
			}
		}
		if (num == -1) {
			result[0] += 1;
		}
		else if (num == 0) {
			result[1] += 1;
		}
		else {
			result[2] += 1;
		}
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        paper = new int[N][N];
        for (int i = 0 ; i < N; i++) {
        	StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        	for (int j = 0 ; j < N; j++) {
        		paper[i][j] = Integer.parseInt(st.nextToken());
        	}
        }
        div(0,0,N);
        System.out.println(result[0]);
        System.out.println(result[1]);
        System.out.println(result[2]);
    }
}